pytorch_lightning/trainer/trainer.py

```diff
@@ -1094,8 +1094,9 @@ def _run_evaluate(self) -> _EVALUATE_OUTPUT:
         # reset trainer on this loop and all child loops in case user connected a custom loop
         self._evaluation_loop.trainer = self
 
-        with self.profiler.profile(f"run_{self.state.stage}_evaluation"), torch.no_grad():
-            eval_loop_results = self._evaluation_loop.run()
+        with self.profiler.profile(f"run_{self.state.stage}_evaluation"):
+            with torch.inference_mode() if hasattr(torch, "inference_mode") else torch.no_grad():
```

Comment:
- Check the PyTorch version explicitly for when inference mode was made available. That should be safer than relying on `hasattr` checks. you could split this out to a separate utility (e.g. `_is_inference_mode_available`) so it can be shared across loops
- Evaluation also includes validation, in addition to testing. Since validation runs during fitting as well, I'm not sure how sae it is to include inference mode if we're fitting. 
To be safe, you could check that the trainer is not fitting before enabling this

Reply:
And let's move this to its own function decorated with `contextmanager` for readability.